File: TestUDP/Assets/Scripts/Data/ServerMessage.cs
public enum ServerMessageType
{
    me, joinedGame, leftGame, positions
}

public class ServerMessage
{
    public ServerMessageType type;
}

public class ServerMessageMe : ServerMessage
{
    public string id;
}

public class ServerMessageJoinedGame : ServerMessage
{
    public PlayerData player;
}

public class ServerMessagePositions : ServerMessage
{
    public PlayerData[] players;
}

public class ServerMessageLeftGame : ServerMessage
{
    public string id;
}

public delegate void OnServerMessageJoinedGameHandler(ServerMessageJoinedGame messageJoinedGame);
public delegate void OnServerMessagePositionsHandler(ServerMessagePositions messagePositions);
public delegate void OnServerMessageLeftGameHandler(ServerMessageLeftGame messageLeftGame);

File: API/data/server_message.ts
import { players } from "../api"
import { Player } from "../components/player"
import { PlayerData } from "./player_data"

enum ServerMessageType {
    me, joinedGame, leftGame, positions
}

class ServerMessage {
    readonly type: ServerMessageType

    protected constructor(type: ServerMessageType) {
        this.type = type
    }
}

class ServerMessageMe extends ServerMessage {
    readonly id: string

    constructor(id: string) {
        super(ServerMessageType.me)
        this.id = id
    }
}

class ServerMessageJoinedGame extends ServerMessage {
    readonly player: PlayerData

    constructor(player: PlayerData) {
        super(ServerMessageType.joinedGame)
        this.player = player
    }
}

class ServerMessagePositions extends ServerMessage {
    readonly players: PlayerData[]

    constructor() {
        super(ServerMessageType.positions)
        this.players = players.getPlayers().map((p: Player) => p.data)
    }
}

class ServerMessageLeftGame extends ServerMessage {
    readonly id: string

    constructor(id: string) {
        super(ServerMessageType.leftGame)
        this.id = id
    }
}

export {
    ServerMessageType, ServerMessage, ServerMessageMe, ServerMessagePositions, ServerMessageJoinedGame, ServerMessageLeftGame
}